source/fuzz/transformation_record_synonymous_constants.cpp

```diff
@@ -19,8 +19,67 @@ namespace spvtools {
 namespace fuzz {
 
 namespace {
-bool IsScalarZeroConstant(const opt::analysis::Constant* constant) {
-  return constant->AsScalarConstant() && constant->IsZero();
+
+// Returns true if the two given constants are equivalent
+// (the description of IsApplicable specifies the conditions they must satisfy
+// to be considered equivalent)
+bool AreEquivalentConstants(opt::IRContext* ir_context, uint32_t constant_id1,
+                            uint32_t constant_id2) {
+  opt::Instruction* def_1 = ir_context->get_def_use_mgr()->GetDef(constant_id1);
+  opt::Instruction* def_2 = ir_context->get_def_use_mgr()->GetDef(constant_id2);
+
+  // Check that the definitions exist
+  assert(def_1 && def_2 && "The constant ids must exist in the module.");
+
+  // The type ids must be the same
+  // TODO(https://github.com/KhronosGroup/SPIRV-Tools/issues/3536): Somehow
+  // relax this for integers (so that unsigned integer and signed integer are
+  // considered the same type)
+  if (def_1->type_id() != def_2->type_id()) {
+    return false;
+  }
+
+  auto constant1 = ir_context->get_constant_mgr()->GetConstantFromInst(def_1);
+  auto constant2 = ir_context->get_constant_mgr()->GetConstantFromInst(def_2);
+
+  assert(constant1 && constant2 && "The ids must refer to constants.");
+
+  // If either constant is null, the other is equivalent iff it is zero-like
+  if (constant1->AsNullConstant()) {
+    return constant2->IsZero();
+  }
+
+  if (constant2->AsNullConstant()) {
+    return constant1->IsZero();
+  }
+
+  // If the constants are scalar, they are equal iff their words are the same
+  if (auto scalar1 = constant1->AsScalarConstant()) {
+    return scalar1->words() == constant2->AsScalarConstant()->words();
+  }
+
+  // If the constants are composite, they are equivalent iff their components
+  // match
+  if (constant1->AsCompositeConstant()) {
+    // Since the types match, we already know that the number of components is
+    // the same. We check that the input operands of the definitions are all
+    // constants and that they are pairwise equivalent.
+    for (uint32_t i = 0; i < def_1->NumInOperands(); i++) {
+      if (!AreEquivalentConstants(ir_context, def_1->GetSingleWordInOperand(i),
+                                  def_2->GetSingleWordInOperand(i))) {
+        return false;
+      }
+    }
+
+    // If we get here, all the components are equivalent
+    return true;
+  }
+
+  // If we get here, the type is not sensible to check for equivalence
+  assert(false &&
+         "Equivalency of constants can only be checked with scalar, composite "
```

Comment:
Equivalency -> Equivalence

Also see comment above about moving this assertion.

Reply:
Done